### PR TITLE
lthash: lthash poc

### PR DESCRIPTION
**Pull request by jumpsiegel**

### CODE REVIEW

**Comment by jumpsiegel** [comment]
This is the first round of lthash changes to match what is being done in the anza tree.   We have tested against the hashes they are generating and it matches in all cases. It also has no statistically significant performance impact.  The lthash is calculated and logged but has no impact on the banks hash or snapshot loading at present.  The next step is to formalize the SIMD given these poc's to break out the progression of feature gates we plan on doing

1. Snapshot hash calculation/checking - since a snapshot is generated often, anza wants this first
2. Epoch Account Hash - since we have not multi threaded this, we want this next 
3. Account Delta Hash - this is last due to the politics of lite clients

File: src/flamenco/runtime/fd_hashes.c [hunk 1]
```diff
@@ -287,17 +262,22 @@ fd_hash_bank( fd_exec_slot_ctx_t * slot_ctx,
...
                   "accounts_delta:   %s\n"
+                  "lthash:           %s\n"
                   "signature_count:  %ld\n"
                   "last_blockhash:   %s\n",
                   slot_ctx->slot_bank.slot,
                   FD_BASE58_ENC_32_ALLOCA( hash->hash ),
                   FD_BASE58_ENC_32_ALLOCA( slot_ctx->prev_banks_hash.hash ),
-                  FD_BASE58_ENC_32_ALLOCA( slot_ctx->account_delta_hash.hash ),
+                  FD_LTHASH_ENC_32_ALLOCA( (fd_lthash_value_t *)fd_type_pun_const( slot_ctx->account_delta_hash.hash ) ),
+                  FD_BASE58_ENC_32_ALLOCA( lhash1 ),
```
Comment (topointon-jump) [inline src/flamenco/runtime/fd_hashes.c:280]: This logging seems wrong

File: src/funk/fd_funk_rec.c [hunk 2]
```diff
@@ -67,7 +68,12 @@ fd_funk_rec_query_global( fd_funk_t *               funk,
...
       fd_funk_xid_key_pair_t pair[1]; fd_funk_xid_key_pair_init( pair, fd_funk_txn_xid( txn ), key );
       fd_funk_rec_t const * rec = fd_funk_rec_map_query_const( rec_map, pair, NULL );
-      if( FD_LIKELY( rec ) ) return rec;
+      if( FD_LIKELY( rec ) ) {
+        if( FD_UNLIKELY(NULL != txn_out  ) ) {
+          *txn_out = txn;
+        }
+        return rec;
+      }
       txn = fd_funk_txn_parent( (fd_funk_txn_t *)txn, txn_map );
```
Comment (topointon-jump) [inline src/funk/fd_funk_rec.c:77]: nit: let's not overwrite txn here, lets make a new variable

File: src/funk/fd_funk_rec.h [hunk 3]
```diff
@@ -175,7 +175,8 @@ fd_funk_rec_query( fd_funk_t *               funk,
 FD_FN_PURE fd_funk_rec_t const *
 fd_funk_rec_query_global( fd_funk_t *               funk,
                           fd_funk_txn_t const *     txn,
-                          fd_funk_rec_key_t const * key );
+                          fd_funk_rec_key_t const * key,
+                          fd_funk_txn_t const **    txn_out );
```
Comment (topointon-jump) [inline src/funk/fd_funk_rec.h:179]: Could we add a doc-comment explaining what txn_out is?

File: src/flamenco/runtime/fd_acc_mgr.h [hunk 4]
```diff
@@ -179,7 +179,8 @@ fd_acc_mgr_view_raw( fd_acc_mgr_t *         acc_mgr,
                      fd_funk_txn_t const *  txn,
                      fd_pubkey_t const *    pubkey,
                      fd_funk_rec_t const ** opt_out_rec,
-                     int *                  opt_err );
+                     int *                  opt_err,
+                     fd_funk_txn_t const ** txn_out   );
```
Comment (topointon-jump) [inline src/flamenco/runtime/fd_acc_mgr.h:183]: Could we add the explanation about txn_out here also? :pray:

File: src/flamenco/runtime/fd_hashes.c [hunk 5]
```diff
@@ -945,31 +958,39 @@ fd_accounts_sorted_subrange( fd_exec_slot_ctx_t * slot_ctx, uint range_idx, uint
...
+      if( fd_acc_exists( metadata ) && memcmp( metadata->hash, &hash, 32 ) != 0 ) {
         FD_LOG_WARNING(( "snapshot hash (%s) doesn't match calculated hash (%s)", FD_BASE58_ENC_32_ALLOCA( metadata->hash ), FD_BASE58_ENC_32_ALLOCA( &hash ) ));
       }
-    }
+    } else
+      fd_memcpy( metadata->hash, &hash, 32 );
 
-    if ((metadata->info.executable & ~1) != 0)
+    if( (metadata->info.executable & ~1) != 0 ) {
+      FD_LOG_ERR((" impossible ")); // remove before merge
```
Comment (topointon-jump) [inline src/flamenco/runtime/fd_hashes.c:981]: nit: remove

File: src/flamenco/runtime/fd_hashes.c [hunk 6]
```diff
@@ -945,31 +958,39 @@ fd_accounts_sorted_subrange( fd_exec_slot_ctx_t * slot_ctx, uint range_idx, uint
...
       continue;
+    }
 
     if( num_pairs == max_pairs ) {
+//      FD_LOG_ERR(("stupid code...  really?")); // remove before merge
```
Comment (topointon-jump) [inline src/flamenco/runtime/fd_hashes.c:986]: nit: remove

File: src/flamenco/runtime/fd_hashes.c [hunk 7]
```diff
@@ -945,31 +958,39 @@ fd_accounts_sorted_subrange( fd_exec_slot_ctx_t * slot_ctx, uint range_idx, uint
...
+    uchar hash[32];
+    fd_lthash_value_t new_lthash_value;
+    fd_lthash_zero(&new_lthash_value);
+
+    if( FD_FEATURE_ACTIVE( slot_ctx, account_hash_ignore_slot ) )
+      fd_hash_account_v1( (uchar *) hash, &new_lthash_value, metadata, rec->pair.key->uc, fd_account_get_data(metadata) );
+    else
+      fd_hash_account_v0( (uchar *) hash, &new_lthash_value, metadata, rec->pair.key->uc, fd_account_get_data(metadata), metadata->slot );
+
+    fd_lthash_add( &accum, &new_lthash_value );
```
Comment (topointon-jump) [inline src/flamenco/runtime/fd_hashes.c:970]: silly question, but why not use `fd_hash_account_current` here?

File: src/ballet/lthash/fd_lthash.h [hunk 8]
```diff
@@ -54,6 +57,69 @@ fd_lthash_sub( fd_lthash_value_t * restrict       r,
...
+    fd_base58_encode_32( _hash, NULL, _y );                              \
+  }                                                                      \
+})
+
+#define FD_LTHASH_ENC_32_ALLOCA( x ) __extension__({                     \
+  char *_out = (char *)fd_alloca_check( 1UL, FD_BASE58_ENCODED_32_SZ );  \
+  FD_LTHASH_ENC_32_BUF(x, _out);                                         \
+  _out;                                                                  \
+})
+
```
Comment (topointon-jump) [inline src/ballet/lthash/fd_lthash.h:91]: Should this call `fd_lthash_fini` and log that? Might make the hashes easier to compare

File: src/flamenco/types/fd_types.json [hunk 9]
```diff
@@ -561,6 +561,15 @@
       ],
       "comment": "https://github.com/anza-xyz/agave/blob/7117ed9653ce19e8b2dea108eff1f3eb6a3378a7/sdk/src/reward_info.rs#L5"
     },
+    {
+      "name": "slot_lthash",
+      "type": "struct",
+      "archival": "true",
+      "alignment": "128",
+      "fields": [
+        { "name": "lthash",                      "type": "uchar[2048]" }
```
Comment (topointon-jump) [inline src/flamenco/types/fd_types.json:570]: Could we add a comment here indicating the type? So that people know to cast it to `fd_lthash_value_t`

Comment (topointon-jump) [reply src/flamenco/types/fd_types.json:570]: Better yet, we could actually make this a `fd_lthash_value_t`?

**Review by topointon-jump — APPROVED** [review]
Looks good! Just some minor comments mostly around documentation/readability.

File: src/ballet/lthash/fd_lthash.h (hunk 8 above)
Comment (jumpsiegel) [reply src/ballet/lthash/fd_lthash.h:91]: This was the shortened hash anza wanted.  Yes, it is not 2k worth of data but this is just to help visually compare we got the same hash

File: src/funk/fd_funk_rec.c (hunk 2 above)
Comment (jumpsiegel) [reply src/funk/fd_funk_rec.c:77]: This was the previous behavior

File: src/flamenco/types/fd_types.json (hunk 9 above)
Comment (jumpsiegel) [reply src/flamenco/types/fd_types.json:570]: This would require changes to the stub generator.  I will add a comment

**Review by topointon-jump — APPROVED** [review]

**Pull request merged**